### PR TITLE
[Impeller] Make porter duff blended atlas calls faster.

### PR DESCRIPTION
**Pull request by jonahwilliams**
This implements part of https://github.com/flutter/flutter/issues/131345

---

Makes all porter-duff draw atlas blending avoid subpass creation, by tweaking the existing porter duff shader to support per-vertex colors and output opacity. This requires a small trick: What the blend shaders and what (drawAtlas/drawVertices) consider the source and destination is reversed. That is a blend mode color filter treats the color as the source whereas drawVertices/drawAtlas treat their per-vertex colors as the destination. Therefore I cannot simply make all color inputs per vertex and use the exact same shaders :(

However, for the porter duff blends we should just be able to reverse the blend mode (srcover -> dstOver, xor->xor) (I'm pretty sure I did this correctly) so that the blending is still correct. I don't think we can do this for the advanced blends, but supporting those is less important than the common atlas blending parameters like dstIn.

This PR only updates drawAtlas.

### CODE REVIEW

File: impeller/entity/contents/filters/blend_filter_contents.cc [hunk 1]
```diff
@@ -456,9 +476,11 @@ std::optional<Entity> BlendFilterContents::CreateForegroundPorterDuffBlend(
     frame_info.texture_sampler_y_coord_scale =
         dst_snapshot->texture->GetYCoordScale();
 
-    frag_info.color = foreground_color.Premultiply();
     frag_info.input_alpha =
         absorb_opacity ? dst_snapshot->opacity * alpha.value_or(1.0) : 1.0;
+    // TODO(jonahwilliams): This filter may not be absorbing opacity correctly,
+    // double check against Skia.
+    frag_info.output_alpha = 1.0;
```
Comment (jonahwilliams) [inline impeller/entity/contents/filters/blend_filter_contents.cc:483]: I think we might be doing this wrong today but I'd need to double checl

File: impeller/tools/malioc.json [hunk 2]
```diff
@@ -8963,28 +8963,27 @@
       "core": "Mali-G78",
       "filename": "flutter/impeller/entity/gles/porter_duff_blend.frag.gles",
       "has_side_effects": false,
-      "has_uniform_computation": true,
+      "has_uniform_computation": false,
```
Comment (jonahwilliams) [inline impeller/tools/malioc.json:8966]: 🤷

File: impeller/entity/contents/filters/blend_filter_contents.cc (hunk 1 above)
Comment (jonahwilliams) [reply impeller/entity/contents/filters/blend_filter_contents.cc:481]: Nah its fine

**Comment by jonahwilliams** [comment]
Friendly ping 😄

File: impeller/tools/malioc.json (hunk 2 above)
Comment (chinmaygarde) [reply impeller/tools/malioc.json:8966]: These moved to the vert.

**Review by chinmaygarde — APPROVED** [review]
@bdero said he would take a look yesterday but he may be preoccupied. I took a stab instead and LGTM. If he has further comments, we can handle it in a later pass. Thanks!

**Review by bdero — APPROVED** [review]
LGTM, but can we add a golden that uses AtlasContents?

Summary of my understanding:
* The `porter_duff_blend` is reworked to distribute the tiles while performing the blend at the same time.
* The blend shader needs to be reversed because the blended colors are actually the destination: https://api.flutter.dev/flutter/dart-ui/Canvas/drawAtlas.html

In an engine sync you mentioned that the blend isn't behaving correctly, but I wasn't able to spot a potential problem.

**Comment by jonahwilliams** [comment]
> In an engine sync you mentioned that the blend isn't behaving correctly, but I wasn't able to spot a potential problem.

I might have mispoke, I didn't mean with this patch but at ToT. In general there is a problem with some of the blending with the current implementation. Those issues are linked in the above issue but in particular: https://github.com/flutter/flutter/issues/118914

So I'm trying to move us closer to what Skia is doing without adding more shaders.

**Comment by bdero** [comment]
Oh! I notice that issue is from January. We discovered some tricky alpha management issues on the advanced blends that needed to be resolved -- maybe it's already fixed by https://github.com/flutter/engine/pull/43283/?

BTW, if you're planning to employ a trick like this to make the advanced blends faster in DrawAtlas and/or DrawVertices, I think that would also be a great to resolve the mask tracking issue I was worried about/get rid of the uncomfortable possibility of edge artifacts.

**Comment by jonahwilliams** [comment]
Oh nice, that's right you fixed that. I'll need to recheck.

**Comment by chinmaygarde** [comment]
Are we landing this?

**Comment by jonahwilliams** [comment]
Waiting on benchmarks

**Comment by jonahwilliams** [comment]
Benchmarks have landed, will monitor and ensure those are running and then land later this week.

**Pull request merged**